core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java

```diff
@@ -303,6 +304,9 @@ public Table create() {
       }
 
       Map<String, String> properties = propertiesBuilder.build();
+      // Validate the metrics
+      MetricsConfig.fromProperties(properties).validateProperties(schema);
```

Comment:
I think we need this in the `createTransaction` and other methods as well.

Reply:
Oh cool, I didn't think of putting it there, I'll give that a shot.

Reply:
@rdblue, will moving the check to the constructor require us to also update the metrics config when the schema evolves? Otherwise, drop/rename operations will always fail if the metrics config refers to them?

Reply:
I think it is a good thing to update the metrics config implicitly when the schema evolves, BTW.

Reply:
I think maybe automatically updating the metrics config on schema evolution could be a good follow up ticket, WDYT @aokolnychyi ?
I think for this first part requiring that the metrics config is valid during schema evolution is a good first step.

Reply:
I agree that it is a follow-up. This will catch the problem and we can rewrite later.

We may also want to consider alternative storage for these that uses ID rather than name. I guess we could rewrite to ID-based.

Reply:
It is absolutely a follow-up item. Just wanted to bring attention to that.

Reply:
Sounds good. I've added https://github.com/apache/iceberg/issues/2079

Reply:
Thanks, @holdenk!

Reply:
After thinking about this a bit more, I don't think it is a good idea to add the check in the `TableMetadata` constructor after all. The problem with doing that is that it would break existing tables because they would validate when loading a table. Instead, we need to validate each case where a table is created so we could put it in the `newTableMetadata` factory method instead.

That would mean we also need to keep the validation in the property and schema update operations as you already have in this PR.